OpenSim/Simulation/Model/GeometryPath.cpp

```diff
@@ -124,44 +121,57 @@ generateDecorations(bool fixed, const ModelDisplayHints& hints,
     if (fixed) { return; }
     // Ensure that the state has been realized to Stage::Dynamics to give
     // clients of this path a chance to calculate meaningful color information.
+    // TODO: this should be removed. Generate decorations is reporting it should 
+    // not force computations.
     this->getModel().getMultibodySystem().realize(state, SimTK::Stage::Dynamics);
 
-    this->updateDisplayPath(state);
-
-    // Even though these are Points which are Components they are completey 
-    // orphaned and not part of any system since they are populated during 
-    // a simulation. TODO we need another data structure to be a DecorativePath
-    // which simply an array of points in ground or on MBs.
-    // Trying to getLocationInGround(state) will faile due to no underlying system.
-    const Array<PathPoint*>& points = getCurrentDisplayPath(state);
-
-    if (points.getSize() == 0) { return; }
+    const Array<PathPoint*>& pathPoints =
+        getCacheVariableValue<Array<PathPoint*> >(state, "current_path");
 
-    const PathPoint* lastPoint = points[0];
+    const PathPoint* lastPoint = pathPoints[0];
     MobilizedBodyIndex mbix(0);
 
     Vec3 lastPos = lastPoint->getLocationInGround(state);
     if (hints.get_show_path_points())
         DefaultGeometry::drawPathPoint(mbix, lastPos, getColor(state), appendToThis);
 
     Vec3 pos;
-    for (int j = 1; j < points.getSize(); j++) {
-        const PathPoint* point = points[j];
-
-        // the body (PhysicalFrame) IS part of the actual Model and its system
-        // so we can ask it for its transform w.r.t. Ground
-        pos = point->getLocationInGround(state);
-
-        if (hints.get_show_path_points())
-            DefaultGeometry::drawPathPoint(mbix, pos, getColor(state), appendToThis);
-
-        
-        // Line segments will be in ground frame
-        appendToThis.push_back(DecorativeLine(lastPos, pos)
-            .setLineThickness(4)
-            .setColor(getColor(state)).setBodyId(0).setIndexOnBody(j));
+    for (int j = 1; j < pathPoints.getSize(); j++) {
+        PathPoint* point = pathPoints[j];
+        PathWrapPoint* mwp = dynamic_cast<PathWrapPoint*>(point);
 
-        lastPos = pos;
+        if (mwp) {
+            // If the point is a PathWrapPoint and has surfacePoints,
```

Comment:
I had copied the code verbatim from the previous implementation in `updateDisplayPath` which was removed. I agree it is cryptic and will attempt to replace with something more intelligible.